src/lib/merkle_ledger/database.ml

```diff
@@ -700,118 +700,88 @@ module Make (Inputs : Inputs_intf) :
+            Direction.map direction ~left:(`Left entry) ~right:(`Right entry)
           in
-          loop directions hashes ((length - 1) :: lengths)
+          loop directions rest_hashes ((length - 1) :: lengths)
             ((dir :: acc_hd) :: acc_tl)
       | _ ->
           failwith "Mismatched lengths"
     in
     loop rev_directions rev_hashes rev_lengths [ [] ]
 
-  let wide_merkle_path_batch mdb locations =
-    let locations =
-      List.map locations ~f:(fun location ->
-          if Location.is_account location then
-            Location.Hash (Location.to_path_exn location)
-          else (
-            assert (Location.is_hash location) ;
-            location ) )
-    in
-    let rev_locations, rev_directions, rev_lengths =
-      let rec loop locations loc_acc dir_acc length_acc =
-        match (locations, length_acc) with
-        | [], _ :: length_acc ->
-            (loc_acc, dir_acc, length_acc)
-        | k :: locations, length :: length_acc ->
-            if Location.height ~ledger_depth:mdb.depth k >= mdb.depth then
-              loop locations loc_acc dir_acc (0 :: length :: length_acc)
-            else
-              let sibling = Location.sibling k in
-              let sibling_dir =
-                Location.last_direction (Location.to_path_exn k)
-              in
-              let loc_acc =
-                match sibling_dir with
-                | Direction.Left ->
-                    sibling :: k :: loc_acc
-                | Direction.Right ->
-                    k :: sibling :: loc_acc
-              in
-              loop
-                (Location.parent k :: locations)
-                loc_acc (sibling_dir :: dir_acc)
-                ((length + 1) :: length_acc)
-        | _ ->
-            assert false
-      in
-      loop locations [] [] [ 0 ]
+  let merkle_path_batch =
+    let update_locs = Fn.compose List.cons Location.sibling in
+    let extract_hashes_exn ~direction:_ hs = (List.hd_exn hs, List.tl_exn hs) in
+    path_batch_impl ~update_locs ~extract_hashes_exn
+
+  let wide_merkle_path_batch =
+    let update_locs k =
+      Fn.compose (List.cons (Location.sibling k)) (List.cons k)
     in
-    let rev_hashes = get_hash_batch_exn mdb rev_locations in
-    let rec loop directions hashes lengths acc =
-      match (directions, hashes, lengths, acc) with
-      | [], [], [], _ (* actually [] *) :: acc_tl ->
-          acc_tl
-      | _, _, 0 :: lengths, _ ->
-          loop directions hashes lengths ([] :: acc)
-      | ( direction :: directions
-        , hash_l :: hash_r :: hashes
-        , length :: lengths
-        , acc_hd :: acc_tl ) ->
-          let dir =
-            Direction.map direction
-              ~left:(`Left (hash_l, hash_r))
-              ~right:(`Right (hash_l, hash_r))
+    let extract_hashes_exn ~direction = function
+      | sibling :: self :: rest ->
+          let el =
+            match direction with
```

Comment:
Here the bug was fixed, version before refactoring had a different order in tuple.

Arguably, refactoring made the concern easier to identify.